scripts/install-build-tools.sh

```diff
@@ -14,35 +15,208 @@ if (( $EUID != 0 )); then
 if [[ "$OSTYPE" == "darwin"* ]]; then
+
+  # macOS does not support running shell scripts as root with homebrew
+  if [ $EUID -eq 0 ]; then
+    echo -e "Mac users should run this script without 'sudo'. Exiting..."
+    exit 1
+  fi
+
   CPUS=$(sysctl -n hw.ncpu)
   # ensure development environment is set correctly for clang
   $SUDO xcode-select -switch /Library/Developer/CommandLineTools
-  brew install llvm@14 googletest google-benchmark lcov make wget cmake curl
+
+  if ! brew --version &>/dev/null; then
+    echo -e "${cyan}Homebrew is required to install dependencies.${end}"
+    exit 1
+  fi
+
+  brew install llvm@14 googletest google-benchmark lcov make wget cmake bash bc
+  brew upgrade bash
+
+  BREW_ROOT=$(brew --prefix)
+
   CLANG_TIDY=/usr/local/bin/clang-tidy
   if [ ! -L "$CLANG_TIDY" ]; then
-    $SUDO ln -s $(brew --prefix)/opt/llvm@14/bin/clang-tidy /usr/local/bin/clang-tidy
+    $SUDO ln -s "${BREW_ROOT}/opt/llvm@14/bin/clang-tidy" /usr/local/bin/clang-tidy
   fi
   GMAKE=/usr/local/bin/gmake
   if [ ! -L "$GMAKE" ]; then
     $SUDO ln -s $(xcode-select -p)/usr/bin/gnumake /usr/local/bin/gmake
   fi
-fi
 
-if [[ "$OSTYPE" == "linux-gnu"* ]]; then
-  $SUDO apt update
-  $SUDO apt install -y build-essential wget cmake libgtest-dev libbenchmark-dev lcov git software-properties-common rsync unzip
+  # check if supported version of python3 is already installed, and save the version
+  for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+    if which python${PY_VERS}; then
+      # save the installed version if found
+      PY_INSTALLED=${PY_VERS}
+      break
+    fi
+  done
+
+  # install valid python version if not installed yet
+  if [ -z "$PY_INSTALLED" ]; then
+    for PY_VERS in "${PYTHON_VERSIONS[@]}"; do
+      FULL_PY="python${PY_VERS}"
+      # try to install python version from homebrew and verify installation
+      if brew install "${FULL_PY}"; then
+        echo "${FULL_PY} installed successfully"
+        PY_INSTALLED=${PY_VERS}
+        break
+      else
+        # don't exit as we can try on another version if not the last one
+        echo "${FULL_PY} installation failed"
+      fi
+    done
+  fi
+
+  # leave if no valid python version is installed
+  if ! $(which python${PY_INSTALLED}); then
+    echo "Python3 install with brew failed, attempted on these versions: ${PYTHON_VERSIONS[@]}"
+    exit 1
+  else
+    # create virtual environment and install python packages for the valid python version
+    PYTHON_PATH=$(which python${PY_INSTALLED})
+    create_venv_install_python "${PYTHON_PATH}" ${PY_INSTALLED}
+  fi
 
-  wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | $SUDO apt-key add -
-  $SUDO add-apt-repository "deb http://apt.llvm.org/focal/ llvm-toolchain-focal-14 main"
+# Linux install with apt
+elif [[ "$OSTYPE" == "linux-gnu"* ]]; then
+  # avoids getting stuck on interactive prompts which is essential for CI/CD
+  export DEBIAN_FRONTEND=noninteractive
+  $SUDO apt update -y
+  $SUDO apt upgrade -y
```

Comment:
Is this required? I don't think we want to enforce apt upgrade on unsuspecting users